core/include/detray/core/detail/tuple_array_container.hpp

```diff
@@ -0,0 +1,306 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Detray include(s)
+#include "detray/core/detail/tuple_container.hpp"
+#include "detray/definitions/qualifiers.hpp"
+
+// Vecmem include(s)
+#include <vecmem/memory/memory_resource.hpp>
+
+// System include(s)
+#include <type_traits>
+
+namespace detray {
+
+/** Tuple based container with array elements.
+ *
+ * @tparam tuple_t is the type of tuple
+ * @tparam array_t is the type of array
+ * @tparam id_t is the type of indexing integer
+ * @tparam Ts... is the types of tuple elements
+ */
+template <template <typename...> class tuple_t,
+          template <typename, std::size_t> class array_t, typename id_t,
+          typename... Ts>
+class tuple_array_container
+    : public tuple_container<tuple_t, id_t,
+                             array_t<typename Ts::object_type, Ts::N>...> {
+
+    public:
+    // Convenient type declarations
+    using base_type =
+        tuple_container<tuple_t, id_t,
+                        array_t<typename Ts::object_type, Ts::N>...>;
+    using base_type::base_type;
+    using id_type = typename base_type::id_type;
+
+    static constexpr std::size_t m_tuple_size = base_type::m_tuple_size;
+
+    template <typename... Args>
+    using tuple_type = typename base_type::template tuple_type<Args...>;
+
+    template <typename T, std::size_t I>
+    using array_type = array_t<T, I>;
+
+    using container_type = typename base_type::container_type;
+    using container_data_type =
+        tuple_type<array_type<typename Ts::data_type, Ts::N>...>;
+    using container_view_type =
+        tuple_type<array_type<typename Ts::view_type, Ts::N>...>;
+
+    /**
+     * Constructor with a vecmem memory resource. (host-side only)
+     *
+     * @param resource is the vecmem memory resource
+     */
+    DETRAY_HOST
+    tuple_array_container(vecmem::memory_resource& resource)
+        : base_type(container_type(to_host<typename Ts::object_type>(
+              resource, std::make_index_sequence<Ts::N>{})...)) {}
```

Comment:
It should be a reference instead of pointer if `Ts` is `grid2`. I was thinking of specializing the constructor for inherited classes. I will do make traits for that

Reply:
Now it receivces a memory resource reference rather pointer which means `vecmem::vector` is not available as an object type anymore. I think it is better because most of detray objects  have a constructor with reference